core/executor/src/heap.rs

```diff
@@ -14,43 +14,446 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
 
-#![warn(missing_docs)]
+//! This module implements a freeing-bump allocator.
+//! See more details at https://github.com/paritytech/substrate/issues/1615.
 
-//! This module implements a linear allocation heap.
+use log::trace;
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+
+// The pointers need to be aligned to 8 bytes.
+const ALIGNMENT: usize = 8;
+
+// The pointer returned by `allocate()` needs to fulfill the alignment
+// requirement. In our case a pointer will always be a multiple of
+// 8, as long as the first pointer is aligned to 8 bytes.
+// This is because all pointers will contain a 8 byte prefix (the list
+// index) and then a subsequent item of 2^x bytes, where x = [3..24].
+const N: usize = 22;
+const MAX_POSSIBLE_ALLOCATION: usize = 16777216; // 2^24 bytes
 
 pub struct Heap {
-	end: u32,
-	total_size: u32,
+	allocated_ptrs: HashMap<usize, bool>,
```

Comment:
this should be removed, or, if needed for debug purposes, put behind a feature gate

Reply:
Yes, I put these into the `Revert me:` commit, so that I can throw them out before merging.